XenAdmin/TabPages/PvsPage.cs

```diff
@@ -0,0 +1,208 @@
+                    dataGridViewVms.Rows[0].Selected = true;
+            }
+            finally
+            {
+                dataGridViewVms.ResumeLayout();
+            }
+        }
+
+        private DataGridViewRow NewPvsFarmRow(PVS_farm pvsFarm)
+        {
+            var farmCell = new DataGridViewTextBoxCell {Value = pvsFarm.name};
+            var configurationCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsFarm.cache_storage.Count > 0 ? Messages.PVS_CACHE_MEMORY_AND_DISK: Messages.PVS_CACHE_MEMORY_ONLY
+            };
+            var cacheSrsCell = new DataGridViewTextBoxCell
+            {
+                Value = string.Join(",  ", Connection.ResolveAll(pvsFarm.cache_storage))
+            };
+
+            var newRow = new DataGridViewRow { Tag = pvsFarm };
+            newRow.Cells.AddRange(farmCell, configurationCell, cacheSrsCell);
+
+            return newRow;
+        }
+
+        private DataGridViewRow NewVmRow(PVS_proxy pvsProxy)
+        {
+            var vm = pvsProxy.VM;
+            System.Diagnostics.Trace.Assert(vm != null);
+            var vmCell = new DataGridViewTextBoxCell {Value = pvsProxy.VM.Name};
+            var cachedCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsProxy.currently_attached ? Messages.YES : Messages.NO
+            };
+            var srCell = new DataGridViewTextBoxCell {Value = Connection.Resolve(pvsProxy.cache_SR)};
+            var prepopulationCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsProxy.prepopulate ? Messages.YES : Messages.NO
+            };
+
+            var newRow = new DataGridViewRow { Tag = vm };
+            newRow.Cells.AddRange(vmCell, cachedCell, srCell, prepopulationCell);
+            vm.PropertyChanged += VmPropertyChanged;
+
+            return newRow;
+        }
+
+        private void UnregisterVMHandlers()
+        {
+            foreach (DataGridViewRow row in dataGridViewVms.Rows)
+            {
+                var vm = row.Tag as VM;
+                if (vm == null) 
+                    continue;
+                vm.PropertyChanged -= VmPropertyChanged;
+            }
+        }
+
+        private void PvsFarmBatchCollectionChanged(object sender, EventArgs e)
+        {
+            Program.Invoke(this, LoadFarms); 
+        }
+
+        private void PvsProxyBatchCollectionChanged(object sender, EventArgs e)
+        {
+            Program.Invoke(this, LoadVMs);
+        }
+
+        private void VmPropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName != "name_label")
+                return;
+
+            foreach (DataGridViewRow row in dataGridViewVms.Rows)
+            {
+                var vm = row.Tag as VM;
+                if (vm != null && vm.Equals(sender))
+                    row.Cells["columnVM"].Value = vm.Name;

```

Comment:
Is there ever only one or zero VMs here which are equal to sender? If so, we could break out of the loop early here once we've found the match.